Picolibc 1.8.5

keith-packard opened the pull request:
Merge in picolibc version 1.8.5.

There is a matching SDK PR pending, https://github.com/zephyrproject-rtos/sdk-ng/pull/707

This offers two new printf variants, 'long long', which provides an integer-only mode with long long support included and a 'minimal' variant which cuts out a bunch of format modifier support to make things very small.

hello-world on cortex-m3 FLASH sizes:
```
float		13378
float inexact	10666
long long	 7950
integer		 7798
minimal		 6834
```
For comparison, here are the minimal C lib sizes:
```
float		11082
long long	 9194
integer		 8426
nano		 7386
```
With this change, all of the cbprintf configuration modes are reflected in the available printf variants, which means the SDK will be able to offer memory savings comparable to the picolibc module.

Closes: #62444 
Closes: #64333

keith-packard committed:
lib/os: With CBPRINTF_NANO, picolibc long-long printf isn't required

CBPRINTF_FULL_INTEGRAL doesn't happen to explicitly conflict with
CBPRINTF_NANO, but when CBPRINTF_NANO is enabled, there's no long long I/O
support provided.

Allow picolibc long-long I/O support to also be elided when CBPRINTF_NANO
is enabled to save similar amounts of space.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/picolibc: Support picolibc's assert-verbose option

This option in picolibc switches the assert macro between a chatty version
and one which provides no information at all. This latter mode avoids
placing the associated strings in memory.

The Zephyr option is PICOLIBC_ASSERT_VERBOSE and it is disable by default.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/picolibc: Support 'long long' and 'minimal' printf variants

Picolibc's 'minimal' printf mode reduces functionality and size even more
than the 'integer' mode. Use this where memory is at a premium and where
the application knows that it does not require exact printf semantics.

1.8.5 adds two more printf variants, 'long long' and 'minimal'. The 'long
long' variant is the same as the 'integer' variant but with long long
support enabled. The 'minimal' variant reduces functionality and size even
more than the 'integer' mode. Applications can use this where memory is at
a premium and where the application does not require exact printf
semantics.

With these two added variants, the SDK has enough options so that all of
the cbprintf modes can be supported with the pre-compiled bits:

 1. CBPRINTF_NANO - picolibc's 'minimal' variant
 2. CBPRINTF_REDUCED_INTEGRAL - picolibc's 'integer' variant
 3. CBPRINTF_FULL_INTEGRAL - picolibc's 'long long' variant
 4. CBPRINTF_FB_SUPPORT - picolibc's 'double' variant

This patch makes the cbprintf Kconfig values drive the default picolibc
variant, disables picolibc variants not capable of supporting the required
cbprintf level, but allows applications to select more functionality in
picolibc than cbprintf requires.

Note that this depends on the SDK including picolibc 1.8.5. Without that,
selecting the 'minimal' or 'long long' variant in Zephyr will end up with
the default variant from picolibc, which is the full version with floating
point support. When using the module things will work as specified.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
west: Update picolibc to version 1.8.5

Picolibc version 1.8.5 offers additional memory savings and new long-long
and minimal printf variants which can be selected from either the SDK or
module version of the library.

This needed a patch to the Zephyr cmake support bits to enable one of the
new picolibc 1.8.5 features.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
net/lwm2m: Provide initialization for path_list_size

The compiler cannot tell that path_list_size is never used without being
set, so help it out by providing a reasonable initialization value.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
This should address issue #62444

keith-packard committed:
doc: Update picolibc section in 3.5 migration guide for 1.8.5

Picolibc 1.8.5 includes more control over printf capabilities, document
those in the migration guide.

Signed-off-by: Keith Packard <keithp@keithp.com>

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_observation.c:
```diff
@@ -529,7 +529,7 @@ struct observe_node *engine_observe_node_discover(sys_slist_t *observe_node_list
 {
 	struct observe_node *obs;
-	int obs_list_size, path_list_size;
+	int obs_list_size, path_list_size = 0;
```
Why is this change?
On what compiler did you see the error/warning code?

keith-packard replied:
running twister using SDK version 0.16.3-rc1 on my local machine. I'm afraid I've lost the precise tree at this point. Looking at the code, I wasn't surprised by the compiler generating a spurious warning here as the assignment and use are both protected inside 'if' clauses. I'd be fine removing this patch, but as I couldn't get a clean twister run locally without it, I included it here assuming it would also be needed for the CI run against the PR.

nordicjm commented:
I can't remember what the smallest flash/RAM device supported in zephyr is but think it's an stm32 chip with something like 2K RAM and 4K flash, maybe @erwango knows? Can we ensure that builds with this library successfully for at least the minimal zephyr application (using picolibc) and, ideally, hello_world using picolibc too with the default app configuration?

erwango commented:
> I can't remember what the smallest flash/RAM device supported in zephyr is but think it's an stm32 chip with something like 2K RAM and 4K flash, maybe @erwango knows?

I don't know if that's Zephyr smallest, but current STM32 Zephyr smallest is nucleo_l011k4: 2K RAM, 16K Flash

keith-packard commented:
> I don't know if that's Zephyr smallest, but current STM32 Zephyr smallest is nucleo_l011k4: 2K RAM, 16K Flash

```
$ west build -b nucleo_l011k4 samples/hello_world/
Memory region         Used Size  Region Size  %age Used
           FLASH:       12886 B        16 KB     78.65%
             RAM:        1356 B         2 KB     66.21%
        IDT_LIST:          0 GB         2 KB      0.00%
$ west build -b nucleo_l011k4 samples/hello_world/ -d build-minimal -- -DCONFIG_MINIMAL_LIBC=y
Memory region         Used Size  Region Size  %age Used
           FLASH:       12962 B        16 KB     79.11%
             RAM:        1328 B         2 KB     64.84%
        IDT_LIST:          0 GB         2 KB      0.00%
$ west build -b nucleo_l011k4 samples/hello_world/ -- -DCONFIG_COMMON_LIBC_MALLOC=n -DCONFIG_PICOLIBC_IO_MINIMAL=y
Memory region         Used Size  Region Size  %age Used
           FLASH:       11842 B        16 KB     72.28%
             RAM:        1344 B         2 KB     65.62%
        IDT_LIST:          0 GB         2 KB      0.00%
```
In the default configuration, it's slightly smaller than the minimal C library version. If you turn off the (unused) malloc code and switch to the smaller printf code, it's even smaller.

keith-packard committed:
tests/kernel: Fix test printk output for new printf variants

In minimal mode, format modifiers are not supported, leading to a lack
of width and precision support.

long long values are presented correctly in either long long or floating
mode.

Signed-off-by: Keith Packard <keithp@keithp.com>

kartben commented:
> This offers a new printf variant which cuts out a bunch of format modifier support while allowing us to add long long support to the 'regular' variant while still offering something smaller.

@keith-packard Could you please try out `samples/subsys/logging/dictionary` code sample?
I've noticed that it seems to be broken with picolibc (but also newlib, IIRC) and I wonder if this PR would by any chance make the situation better and/or be an opportunity to address other outstanding issues around format modifiers. Thank you!

keith-packard commented:
> @keith-packard Could you please try out `samples/subsys/logging/dictionary` code sample? I've noticed that it seems to be broken with picolibc (but also newlib, IIRC) and I wonder if this PR would by any chance make the situation better and/or be an opportunity to address other outstanding issues around format modifiers. Thank you!

The python log parser code was busted and didn't handle signed char/unsigned char length modifiers for integer specifiers. #63007

keith-packard commented on tests/kernel/common/src/printk.c:
```diff
@@ -17,6 +17,27 @@ int (*_old_char_out)(int);
 
 #if defined(CONFIG_PICOLIBC)
 
+#define ZEPHYR_PICOLIBC_VERSION (__PICOLIBC__ * 10000 + \
```
So just skip tests in that case instead of verifying that they generate the expected result? I'm not sure that's as useful for this test where we're kinda walking through C library configurations and making sure they work as intended?

cfriedt replied:
Output is more or less an unexpected result with minimal io config anyway.

I guess the question is how specific do we want to be about the deviation. Fine to be overly specific as well, but just saying you have the option to be less specific too, which means PicoLibC would not be contractually bound to Zephyr to ensure minimal io formatting is a certain way. I.e. if there ever need to be modifications to picolibc's minimal io formatter, to save a few byte here or there, there is more flexibility.

keith-packard replied:
Yeah, that's a fair point. I think I'll leave this patch and burn that bridge when we get to it.

keith-packard commented:
As you may have noticed, this PR has changed from pulling in picolibc 1.8.4 to picolibc 1.8.5 -- switching 1.8.4 to enable long long for integer mode increased the size of many applications. 1.8.5 has a new 'long long' variant which allows applications to easily select between those two modes to save space. With this, the SDK will also support separate integer and long long variants.

keith-packard committed:
samples/code_relocation_nocopy: Increase fake flash size

When linking with the 0.16.3 SDK version of picolibc, using long long
cbprintf support pulls in the full floating point printf which is much
larger than the integer-only version. This ends up overflowing the memory
region available for it. Increase the size of that by bumping the start of
the fake region by 8kB.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
For this to get merged in 3.5.0, we'll need a new SDK release including picolibc 1.8.5 as well. If that isn't going to happen, then this needs to wait.

jhedberg commented:
> For this to get merged in 3.5.0, we'll need a new SDK release including picolibc 1.8.5 as well. If that isn't going to happen, then this needs to wait.

I had a short chat with Anas about this. It's inevitably post 3.5 stuff.

keith-packard commented:
> > For this to get merged in 3.5.0, we'll need a new SDK release including picolibc 1.8.5 as well. If that isn't going to happen, then this needs to wait.
> 
> I had a short chat with Anas about this. It's inevitably post 3.5 stuff.

That's what I expected. No good reason to make this large a change with only a few days to go.

keith-packard commented:
I'm seeing the same failure on main. This test is built with the external C library by default, so I'm not sure what updating picolibc would do anyways...

keith-packard commented:
With SDK 0.16.4 released, this should be ready to merge. I've rebased to re-test against current bits.

mniestroj reviewed (approved):
LGTM.

I can confirm, that #62444 is solved with this PR.
